java/server/src/org/openqa/selenium/grid/distributor/local/LocalDistributor.java

```diff
@@ -88,33 +106,122 @@
+        fireSessionRejectedEvent("New session request timed out", reqId);
+      } else {
+        try {
+          CreateSessionResponse sessionResponse = newSession(sessionRequest);
+          NewSessionResponse newSessionResponse =
+              new NewSessionResponse(reqId, sessionResponse.getSession(),
+                                     sessionResponse.getDownstreamEncodedResponse());
+
+          bus.fire(new NewSessionResponseEvent(newSessionResponse));
+        } catch (SessionNotCreatedException e) {
+          // If error is due to the slots being busy, adding to front of queue else reject the request
+          // If that fails, then just reject the request.
+          if (e.getMessage().startsWith("Unable to find provider for session") ||
+              e.getMessage().startsWith("Unable to reserve a slot for session request")) {
+            boolean retried = this.sessionRequests.retryAddToQueue(sessionRequest, reqId);
+
+            attributeMap.put("request.retry_add", EventAttribute.setValue(retried));
+            span.addEvent("Retry adding to front of queue. All slots are busy.", attributeMap);
+
+            if (!retried) {
+              span.addEvent("Retry adding to front of queue failed.", attributeMap);
+              fireSessionRejectedEvent(e.getMessage(), reqId);
+            }
+          } else {
+            fireSessionRejectedEvent(e.getMessage(), reqId);
+          }
+        }
+      }
+    } finally {
+      span.close();
+    }
+  }
+
+  private boolean hasRequestTimedOut(HttpRequest request) {
+    String enqueTimestampStr = request.getHeader(NewSessionQueue.SESSIONREQUEST_TIMESTAMP_HEADER);
+    Instant enque = Instant.ofEpochSecond(Long.parseLong(enqueTimestampStr));
+    Instant deque = Instant.now();
+    Duration duration = Duration.between(enque, deque);
+
+    return duration.compareTo(sessionRequestTimeout) > 0;
+  }
+
+  private void fireSessionRejectedEvent(String message, RequestId reqId) {
+    bus.fire(
+        new NewSessionRejectedEvent(new NewSessionErrorResponse(reqId, message)));
   }
 
   public static Distributor create(Config config) {
     Tracer tracer = new LoggingOptions(config).getTracer();
     EventBus bus = new EventBusOptions(config).getEventBus();
     HttpClient.Factory clientFactory = new NetworkOptions(config).getHttpClientFactory(tracer);
     SessionMap sessions = new SessionMapOptions(config).getSessionMap();
+    NewSessionQueuer sessionRequests =
+        new NewSessionQueuerOptions(config).getSessionQueuer(
+            "org.openqa.selenium.grid.sessionqueue.remote.RemoteNewSessionQueuer");
+    Duration requestTimeout = new NewSessionQueueOptions(config).getSessionRequestTimeout();
     BaseServerOptions serverOptions = new BaseServerOptions(config);
 
-    return new LocalDistributor(tracer, bus, clientFactory, sessions, serverOptions.getRegistrationSecret());
+    return new LocalDistributor(tracer,
```

Comment:
Style nit: indentation. It's better with these kinds of things to move the `tracer` to a new line, and then everything lines up nicely.